fix: incorrect region access with split

haraldh opened the pull request:
For `unmap`: instead of inheriting the access of the previous record, the split records gained the default/empty access.
Also check `unmap` with `Access::READ`. This would have failed without the `unmap` fix.

And some minor patches:

* Make Access generic, so users of the crate can define their
    own Access flags.
* fix: add SPDX-License-Identifier  
* fix debug output
* add mixed access ledger test
* add comments and early returns

haraldh committed:
fix: add SPDX-License-Identifier

Signed-off-by: Harald Hoyer <harald@profian.com>

haraldh committed:
feat: generic Access

Make Access generic, so users of the crate can define their
own Access flags.

Signed-off-by: Harald Hoyer <harald@profian.com>

haraldh committed:
test: add mixed access ledger test

Signed-off-by: Harald Hoyer <harald@profian.com>

haraldh committed:
fix: incorrect region access with split

Instead of inheriting the access of the previous record,
the split records gained the default/empty access.

Signed-off-by: Harald Hoyer <harald@profian.com>

haraldh committed:
feat: add comments and early returns

Signed-off-by: Harald Hoyer <harald@profian.com>

haraldh committed:
tests: fix debug output

Showed the access of the wrong record.

Signed-off-by: Harald Hoyer <harald@profian.com>

haraldh committed:
test: check `unmap` with `Access::READ`

This would have failed without the previous `unmap` fix.

Signed-off-by: Harald Hoyer <harald@profian.com>

bstrie commented on src/lib.rs:
```diff
@@ -136,23 +103,19 @@ impl<const N: usize> Ledger<N> {
 
     /// Create a new instance. Make sure that `region.start < region.end`,
     /// because otherwise the ledger will exhibit undefined behaviour.
```
I'm concerned at this comment. If the caller of this function needs to uphold that region.start < region.end or else UB will result, then should this be an `unsafe fn`?

haraldh replied:
it's not `unsafe` and UB in the rust sense... but I fixed it.

bstrie commented on src/lib.rs:
```diff
@@ -394,34 +357,85 @@ mod tests {
+            /// Execute access
+            const EXECUTE = 1 << 2;
+
+            const MMAPED = 1 << 3;
```
What is this? Should it be spelled `MMAPPED`?

haraldh replied:
removed

bstrie commented on src/lib.rs:
```diff
@@ -394,34 +357,85 @@ mod tests {
+                    'x'
+                } else {
+                    '-'
+                }
```
What about the new MMAPED flag?

haraldh replied:
removed

bstrie commented on src/lib.rs:
```diff
@@ -667,6 +691,47 @@ mod tests {
+                region: Region::new(Address::new(record.0 << 12), Address::new(record.1 << 12)),
+                access: record.2,
+            })
+            .collect::<Vec<_>>();
```
The two statements above look like boilerplate shared among all the tests, do you think they should be split out into `to_regions` and `to_records` functions?

haraldh replied:
done

bstrie commented on src/lib.rs:
```diff
@@ -667,6 +691,47 @@ mod tests {
+            .collect::<Vec<_>>();
+
+        let mut ledger = MIXED_LEDGER.clone();
+        assert_eq!(ledger.records(), &[LOWER_HALF_R, UPPER_HALF_W]);
```
Is this assert necessary?

haraldh replied:
not really.. was just a blind copy&paste of the other test

haraldh replied:
removed

bstrie commented on src/lib.rs:
```diff
@@ -314,20 +314,29 @@ impl<T: LedgerAccess, const N: usize> Ledger<T, N> {
-            if region.end < record_start || region.start > record_end {
+            if region.start >= record_end {
                 // Skip:
+                // [   ]   XXXXX
```
These comments are cute, but they don't explain why some of these cases cause `continue` while others cause `return` (which is what I find confusing). In general I find the control flow in this function confusing, can some of these ifs/continues be condensed into if-else?

haraldh replied:
added more

haraldh replied:
and change the control flow to a match statement, so we can catch the `unreachable!()` on logic error.

haraldh committed:
test: factor out setup functions

Signed-off-by: Harald Hoyer <harald@profian.com>

haraldh committed:
fix: add Span and fix `new()` signature

By using `start` + `length`, the problem of checking `end > start` is avoided.
By using `Span`, we don't have to think about constructing the `Region`.

Signed-off-by: Harald Hoyer <harald@profian.com>

haraldh committed:
fix(unmap): use match statement for region checking

and use `unreachable!` for the cases we deem not to be reachable.

This prevents accidentally falling to a case, which was not intended.

Signed-off-by: Harald Hoyer <harald@profian.com>
